### ISSUE TITLE
Typing

### ISSUE THREAD

**Issue by JeppeKlitgaard** [argaen/aiocache#512]
Since `aiocache` only supports modern Python 3 versions, I was wondering if there would be an interesting in having the library be typed?

I wouldn't mind doing the work to get this done, but wanted to get feedback on whether this is something that the maintainers of `aiocache` would even want.

On that same note, how do people feel about dropping Python 3.5 support now that the release has officially reached end-of-life.

**Comment by argaen** [comment]
Hey @JeppeKlitgaard, I don't have anything against adding typing so go for it 👍. Same with dropping 3.5 support, been a while since wanted to do this plus adding support for 3.8 (and 3.9 now) but if you want to give it a try, please do :)

**Comment by Dreamsorcerer** [comment]
Mypy has been configured on the repository, so if anyone wants to start contributing to this, please feel free to make a PR that adds some more annotations.

To work on it, start by uncommenting `check_untyped_defs` in .mypy.ini and fixing the errors that appear when running `mypy`.
If there are too many errors to start with, you can disable the option in sections of the code initially until it is manageable. e.g.
```
[mypy-tests.*]
check_untyped_defs = False
```

**Comment by Olegt0rr** [comment]
@Dreamsorcerer is there an option to split this issue?

1) add paramspec to the decorator 
2) add other typing and mypy checks

Looks like 1 has no blocker and can be published fast

Just added some strings for example and got expected behaviour:
<img width="324" alt="Screenshot 2024-09-02 at 23 47 13" src="https://github.com/user-attachments/assets/221aa5d8-de05-40f2-9dbe-6571a3851dd8">

**Comment by Dreamsorcerer** [comment]
No, exposing type hints at all causes type errors in everyone's code with v0. I tried enabling it before, but had to disable it again.

The v1 milestone is listed in the order that tasks should be completed:
https://github.com/aio-libs/aiocache/milestone/8

Once mypy is running, we could potentially start taking typing contributions, though the other tasks listed before this one will ensure the easiest path to typing.

**Comment by Olegt0rr** [comment]
Even single TypeVar? 
It looks fully abstract and compatible with any decorated functions

Just tried to run mypy with my example and it passed well

P.S.: I looked at v1 - there is a fairly large amount of improvements. I definitely won't be able to switch on quickly. At the same time, I would like to make the current version at least a little more convenient

**Comment by Dreamsorcerer** [comment]
> Even single TypeVar?

Even without a single TypeVar. Just adding the py.typed file resulted in type errors for users, got complaints the moment I pushed a release with it in.

**Comment by Kilo59** [comment]
@Dreamsorcerer @Olegt0rr 

>Even without a single TypeVar. Just adding the py.typed file resulted in type errors for users, got complaints the moment I pushed a release with it in.

That's only true if you add `py.typed` to the top-level package.
It's possible to selectively export types by putting these files in specific subpackages. 

If we want to achieve this for `decorators.py` it would have to be refactored into a subpackage, and you'd need to remove the top-level `py.typed` file that exists right now.
```
aiocache
├── __init__.py
├── backends
│   ├── __init__.py
│   ├── memcached.py
│   ├── memory.py
│   └── redis.py
├── base.py
├── decorators
│   ├── __init__.py
│   ├── foobar.py
│   ├── py.typed
├── exceptions.py
├── factory.py
├── lock.py
├── plugins.py
└── serializers
    ├── __init__.py
    └── serializers.py
```



## PEP-0561

https://peps.python.org/pep-0561/#packaging-type-information

## Example from `great_expectations`

- https://github.com/great-expectations/great_expectations/pull/7438
